Рандомная фигня

TheAirBlow opened the pull request:
- [x] Использовать shifting для нормальной передачи `uint16_t`
- [ ] Перепис приложения Processing, закомментировать все
- [x] Закомментировать все в прошивке, изменить названия переменных и методов
- [x] Ответный пакет для тех, которые таковых не имеют
- [x] Использовать submodules для библиотек
- [x] Документация в `markdown`
- [x] Heartbeat пакет

Processing 3.5.4
Arduino IDE 1.8.19

TheAirBlow committed:
Readme: Update TODO

TheAirBlow committed:
Readme: Add link to ESP8266

TheAirBlow committed:
Readme: Delete TODO, add Windows/Linux link

TheAirBlow committed:
Misc: Delete Android directory

TheAirBlow committed:
Libraries: Use submodules, information text

TheAirBlow committed:
Documentation: Now in markdown, more informative

TheAirBlow commented:
@AlexGyver, что это делает?
```c
delay(currentIP[3] * 2);
answ[0] = 0;
answ[1] = currentIP[3];
reply(answ, 2);
```

Почему он ждет 1 миллисекунду (`192.168.4.1`, к примеру), и отправляет только 4 цифру из IP?

`String ip = brIP.substring(0, brIP.lastIndexOf('.')+1) + str(data[1]);`, что за фокусы?

P.S. Понял. Он последную цифру заменяет на полученную. Если он пакет отправил, значит с IP все норм, смысл с ним фокусы творить?

GyverLibs commented:
1. Он ждёт не 1 мс, а удвоенное количество своего четвертого октета, чтобы избежать коллизий при нескольких гирляндах в сети. Отправляет только четвертый октет адреса, потому что первые три приложение и так знает
2. Этот фокус заменяет четвертый октет адреса смартфона на полученный от гирлянды, делая таким образом её адрес, вроде в комментариях не нуждается

TheAirBlow commented:
> 1. Он ждёт не 1 мс, а удвоенное количество своего четвертого октета, чтобы избежать коллизий при нескольких лампах в сети. Отправляет только четвертый октет адреса, потому что первые три приложение и так знает
> 
>     2. Этот фокус заменяет четвертый октет адреса смартфона на полученный от гирлянды, делая таким образом её адрес, вроде в комментариях не нуждается

Заметил, он на 4 умножает

TheAirBlow commented:
Он успешно отправил пакет и получил ответ, значит IP правильный и в исправлениях не нуждается.
Вроде логично, или я что-то тут не понимаю?

А. Он часть IP брутфорсит. Теперь понял.

GyverLibs commented:
Это не исправление. Мы запрашиваем у гирлянды её адрес и его записываем себе

TheAirBlow commented:
> Это не исправление. Мы запрашиваем у гирлянды её адрес и его записываем себе

Уже понял. Не проще не перебирать IP, а просто получить все из текущей сети?

GyverLibs commented:
Ничего он не брутфорсит. Приложение находится в той же локальной сети, что и есп. По широковещательному ip отправляется команда поиска, и каждая есп отправляет свой 4-й октет

GyverLibs commented:
А где перебор то я не понимаю?

TheAirBlow commented:
Мне надо лучше в Processing разбораться, ничего не понимаю, непривычно)
Говорю фигню которая взбрела в голову случайно

GyverLibs commented:
Это обычная java, по сути тот же срр если не вникать в ооп и прочее чего нет в гувертвинк

TheAirBlow commented:
> Ничего он не брутфорсит. Приложение находится в той же локальной сети, что и есп. По широковещательному IP отправляется команда поиска, и каждая есп отправляет свой 4-й октет

Самое первое о чем бы я подумал - это просто достать все IP из текущей сети, про широковещальный IP просто бы не додумался.

GyverLibs commented:
Есть более элегантные способы найти еспхи, но было абсолютно некогда их исследовать)))

TheAirBlow commented:
> Есть более элегантные способы найти еспхи, но было абсолютно некогда их исследовать)))

Ладно, спасибо что мою голову освежил.
Щас коммитну последние изменения

TheAirBlow committed:
Multiple changes: In description

1) Documentation: Updated and fixed IP requiest
2) Firmware: Added a lot of comments, renamed variables and methods
3) Misc: FastLED as a submodule

TheAirBlow commented:
Ай, ошибка в сообщении коммита

TheAirBlow committed:
Firmware: Heartbeat packet

TheAirBlow committed:
Multiple changes: In description

1) Documentation: Replaced "n / 100", "n % 100" and similar
2) Firmware: Fixed uint16_t read

heX16 commented:
> Самое первое о чем бы я подумал - это просто достать все IP из текущей сети, про широковещальный IP просто бы не додумался.

"просто достать" не возможно, только брутфорсом искать. Поэтому нужен широковещальный пакет, это стандартная практика.

ObjatieGroba commented:
Я предлагаю на pages.github.com отдавать статикой страницу с JS, который будет обнаружать устройство и переходить туда.

Веб сервер на самом ардуино уже есть в соседнем MR. 

Изучаю вопрос об сканировании веб камерой прямо с браузера. То что нашел работает только с компа, но не работает с телефонов. 

Предлагаю также переписать эффекты. Текущие все плоские. Надо елку представить как прямоугольную склеенную ленту.

Текущее сканирование часто косячит - особенно когда не все светодиоды видны из этой точки доступа. Тут  можно постпроцессингом чуток понять что светодиоды кривые и поправить (те что были при сканировании очень тусклыми по сравнению с другими заменить на среднее положение от соседей). Или разрешить нарисовать область, где светодиодов быть не может - чтобы руками пользователь отметал всякие зеркальные поверхности (как у меня от пола отсвечивает)

GyverLibs commented:
Я думаю нужно отказаться от прямого сканирования камерой, это было сделано по приколу для видео. Гирлянду нужно мотать ровной спиралью и камерой именно определять количество витков, либо задавать его просто руками исходя из спиральности расположения

ObjatieGroba commented:
> Гирлянду нужно мотать ровной спиралью и камерой именно определять количество витков

Это не будет работать. Потому что не будет никогда.
Сейчас в этом и прикол что выглядит сцена очень приятно глазу. Ровных елок не бывает, не бывает и ровных мотков. А без это будет обычная гирлянда, коих мильон

GyverLibs commented:
Это как раз таки будет отлично работать, потому что гирлянда мотается на горизонтальный "этаж" ёлки и однократно переходит на следующий. Если соблюдать это правило намотки, позиции всех светодиодов будут понятны с достаточной точностью для вывода простых эффектов, таких как в видео. И это будет точно работать лучше со всех сторон ёлки, чем в текущей плоской реализации

ObjatieGroba commented:
>  горизонтальный "этаж" ёлки и однократно переходит на следующий

Как задавать количество пикселей на этаже?

Как в таком случае будет работать диагональный эффект? Кажется это будет сильно менее ровно чем сейчас

Почему люди захотят отказаться от старого решения в пользу нового в таком случае если прошлое требовало меньше усилий?

Я попытаюсь сделать нормальное сканирование, но не обещаю что в ближайшее время. В целом проблемы и со стабилизацией и с трехмерностью решаемые. И даже с браузера айфона все будет работать  и не требовать какой либо численной настройки от пользователя

GyverLibs commented:
- Да разные способы можно придумать, начиная от ползунка, зажигающего все леды по порядку и создания списка на каком леде будет переход
- Замечательно будет работать
- Потому что усилия нужно приложить один раз за сезон и работать оно будет заведомо лучше. Да и усилиями это сложно назвать, стоять две минуты неподвижно с телефоном гораздо сложнее
